File: desktop/src/lib/escpos.ts
import type { StockMovementBatchDetail, TransactionDetail } from "./types";
import { paperCols, type CashDrawerPin, type ReceiptConfig } from "./receipt";

const formatQty = (n: number) => (Number.isInteger(n) ? n.toString() : n.toFixed(2));

const ESC = 0x1b;
const GS = 0x1d;

function money(n: number): string {
  return "Rp" + Math.round(n || 0).toLocaleString("id-ID");
}

/** Printer thermal umumnya CP437/ASCII, bukan UTF-8 — buang diakritik, ganti sisanya dengan '?'. */
function asciiBytes(s: string): number[] {
  const norm = s.normalize("NFKD").replace(/[̀-ͯ]/g, "");
  const out: number[] = [];
  for (const ch of norm) {
    const code = ch.codePointAt(0) ?? 63;
    out.push(code < 128 ? code : 63);
  }
  return out;
}

class EscPosBuilder {
  private bytes: number[] = [];
  raw(...b: number[]): this {
    this.bytes.push(...b);
    return this;
  }
  text(s: string): this {
    this.bytes.push(...asciiBytes(s));
    return this;
  }
  line(s = ""): this {
    return this.text(s).raw(0x0a);
  }
  init(): this {
    return this.raw(ESC, 0x40);
  }
  align(a: "left" | "center" | "right"): this {
    const n = a === "center" ? 1 : a === "right" ? 2 : 0;
    return this.raw(ESC, 0x61, n);
  }
  bold(on: boolean): this {
    return this.raw(ESC, 0x45, on ? 1 : 0);
  }
  /** Double-height text (GS ! n). Makes text taller while keeping normal width. */
  doubleHeight(on: boolean): this {
    return this.raw(GS, 0x21, on ? 0x01 : 0x00);
  }
  feed(lines: number): this {
    return this.raw(ESC, 0x64, Math.max(0, Math.min(255, lines)));
  }
  /** Potong kertas. `partial=true` = potong sebagian (umum untuk struk beruntun). */
  cut(partial = true): this {
    return this.raw(GS, 0x56, partial ? 1 : 0);
  }
  build(): Uint8Array {
    return new Uint8Array(this.bytes);
  }
}

/**
 * Byte pembuka laci kasir (ESC p m t1 t2 — "drawer kick"). Laci tidak
 * tersambung ke PC, tapi ke port RJ11 di belakang printer, jadi perintahnya
 * dikirim lewat jalur cetak yang sama.
 *
 * `m` menentukan pin konektor: 0 = pin 2 (paling umum), 1 = pin 5. t1/t2 =
 * lama pulsa on/off dalam kelipatan 2 ms (25 → 50 ms, 250 → 500 ms) — nilai
 * aman yang dipakai kebanyakan laci; terlalu pendek bisa bikin solenoid tidak
 * kuat menarik. Sengaja BUKAN bagian dari buildReceiptEscPos supaya bisa
 * dipakai sendiri untuk tombol "Buka Laci" manual.
 */
export function buildDrawerKick(drawer: CashDrawerPin): Uint8Array {
  const pulse = (m: number) => [ESC, 0x70, m, 0x19, 0xfa];
  switch (drawer) {
    case "pin2": return new Uint8Array(pulse(0));
    case "pin5": return new Uint8Array(pulse(1));
    case "both": return new Uint8Array([...pulse(0), ...pulse(1)]);
    default: return new Uint8Array();
  }
}

/**
 * Sisipkan perintah buka laci di AKHIR job cetak (sesudah feed + potong), jadi
 * satu kali kirim ke printer. Sengaja di akhir, bukan di awal: struk diawali
 * `ESC @` (reset printer), dan menaruh perintah laci sebelum reset itu
 * mengandalkan perilaku printer yang tidak seragam antar merek.
 */
export function withDrawerKick(bytes: Uint8Array, drawer: CashDrawerPin): Uint8Array {
  const kick = buildDrawerKick(drawer);
  if (kick.length === 0) return bytes;
  const out = new Uint8Array(bytes.length + kick.length);
  out.set(bytes, 0);
  out.set(kick, bytes.length);
  return out;
}

/** Bangun urutan byte ESC/POS untuk struk transaksi, diakhiri feed + autocut. */
export function buildReceiptEscPos(detail: TransactionDetail, cfg: ReceiptConfig): Uint8Array {
  const w = paperCols(cfg.paper);
  const line = (ch = "-") => ch.repeat(w);
  const center = (t: string) => {
    const s = t.slice(0, w);
    const pad = Math.max(0, Math.floor((w - s.length) / 2));
    return " ".repeat(pad) + s;
  };
  const two = (l: string, r: string) => {
    const left = l.slice(0, Math.max(0, w - r.length - 1));
    const space = Math.max(1, w - left.length - r.length);
    return left + " ".repeat(space) + r;
  };

  const { show } = cfg;
  const b = new EscPosBuilder();
  b.init();
  b.align("center");

  // Header — printer is already in center-align mode (ESC a 1), so do NOT
  // also pad with center(); that double-centres the text (shifts it right).
  if (show.storeName) b.bold(true).doubleHeight(true).line(cfg.storeName).doubleHeight(false).bold(false);
  if (show.address && cfg.address.trim()) cfg.address.split("\n").map((x) => x.trim()).filter(Boolean).forEach((a) => b.line(a));
  if (show.phone && cfg.phone.trim()) b.line(cfg.phone.trim());
  if (show.taxId && cfg.taxId.trim()) b.line(`NPWP: ${cfg.taxId.trim()}`);
  if (show.social) {
    const social = [
      cfg.instagram.trim() && `IG: ${cfg.instagram.trim()}`,
      cfg.tiktok.trim() && `TikTok: ${cfg.tiktok.trim()}`,
      cfg.whatsapp.trim() && `WA: ${cfg.whatsapp.trim()}`,
    ]
      .filter(Boolean)
      .join(" · ");
    if (social) b.line(social);
  }
  if (show.header) {
    cfg.header
      .split("\n")
      .map((x) => x.trim())
      .filter(Boolean)
      .forEach((h) => b.line(h));
  }
  if (show.date) b.line(new Date(detail.created_at).toLocaleString("id-ID"));
  if (show.invoiceNo) b.line(detail.invoice_no);

  if (show.items) {
    b.align("left").line(line());
    for (const it of detail.items) {
      b.line(it.name.slice(0, w));
      b.line(two(`  ${it.qty} x ${money(it.price)}`, money(it.price * it.qty)));
      if (it.discount > 0) b.line(two("  Diskon", "-" + money(it.discount)));
    }
  }

  const hasSummary =
    show.totalItem || show.subtotal || show.discount || show.total || show.paymentMethod || show.change;
  if (hasSummary) {
    b.align("left").line(line());
    // Total item = jumlah semua pcs yang dibeli (bukan jumlah jenis barang).
    if (show.totalItem) {
      b.line(two("Total Item", formatQty(detail.items.reduce((s, it) => s + it.qty, 0))));
    }
    if (show.subtotal) b.line(two("Subtotal", money(detail.subtotal)));
    if (show.discount) b.line(two("Diskon", "-" + money(detail.discount)));
    if (show.total) b.bold(true).line(two("TOTAL", money(detail.total))).bold(false);
    if (show.paymentMethod) {
      if (detail.payment_method === "Kombinasi") {
        b.line(two("Tunai", money(detail.paid_cash ?? 0)));
        b.line(two("QRIS", money(detail.paid_qris ?? 0)));
      } else {
        b.line(two(detail.payment_method, money(detail.paid)));
      }
    }
    if (show.change) b.line(two("Kembali", money(detail.change)));
    b.line(line());
  }

  if (show.footer && cfg.footer.trim()) {
    b.align("center");
    // Footer — same as header: printer is in center-align, so no manual padding.
    cfg.footer
      .split("\n")
      .map((x) => x.trim())
      .filter(Boolean)
      .forEach((f) => b.line(f));
  }

  b.feed(3);
  b.cut(true);

  return b.build();
}

/** Bangun urutan byte ESC/POS untuk dokumen Item Masuk/Keluar, diakhiri feed + autocut. */
export function buildStockDocEscPos(detail: StockMovementBatchDetail, cfg: ReceiptConfig): Uint8Array {
  const w = paperCols(cfg.paper);
  const line = (ch = "-") => ch.repeat(w);
  const center = (t: string) => {
    const s = t.slice(0, w);
    const pad = Math.max(0, Math.floor((w - s.length) / 2));
    return " ".repeat(pad) + s;
  };
  const two = (l: string, r: string) => {
    const left = l.slice(0, Math.max(0, w - r.length - 1));
    const space = Math.max(1, w - left.length - r.length);
    return left + " ".repeat(space) + r;
  };

  const verb = detail.kind === "in" ? "ITEM MASUK" : "ITEM KELUAR";
  const { show } = cfg;
  const b = new EscPosBuilder();
  b.init();
  b.align("center");

  if (show.storeName && cfg.storeName.trim()) b.bold(true).doubleHeight(true).line(cfg.storeName).doubleHeight(false).bold(false);
  b.bold(true).line(verb).bold(false);
  b.line(detail.no);
  b.line(new Date(detail.created_at).toLocaleString("id-ID"));

  b.align("left").line(line());
  for (const it of detail.items) {
    b.line(it.product_name.slice(0, w));
    b.line(two(`  Qty: ${formatQty(it.qty)}`, ""));
    if (it.note) b.line(`  Ket: ${it.note}`.slice(0, w));
  }
  b.line(line());
  b.line(two("Total Item", String(detail.items.length)));
  b.line(two("Total Qty", formatQty(detail.items.reduce((s, it) => s + it.qty, 0))));
  if (detail.note) b.line(`Catatan: ${detail.note}`.slice(0, w));
  if (detail.user_id) b.line(`Oleh: ${detail.user_id}`.slice(0, w));
  b.line(line());

  if (show.footer && cfg.footer.trim()) {
    b.align("center");
    cfg.footer
      .split("\n")
      .map((x) => x.trim())
      .filter(Boolean)
      .forEach((f) => b.line(f));
  }

  b.feed(2);
  b.cut(true);

  return b.build();
}

/** Satu baris data pada tabel/ringkasan laporan yang mau dicetak sebagai struk. */
export interface ReportEscPosRow {
  cells: string[];
  bold?: boolean;
}

/** Satu kartu/tabel dalam laporan (mis. "Per Barang", ringkasan pembayaran). */
export interface ReportEscPosSection {
  heading?: string;
  /** Nama kolom (dari <thead>). Kosong/undefined untuk tabel label/nilai (2 kolom, tanpa header). */
  columns?: string[];
  rows: ReportEscPosRow[];
}

export interface ReportEscPosDoc {
  title: string;
  subtitle?: string;
  meta?: string;
  sections: ReportEscPosSection[];
}

/**
 * Bangun urutan byte ESC/POS untuk laporan generik (dipetakan dari tabel HTML
 * di layar) yang dicetak langsung ke printer thermal, diakhiri feed + autocut.
 */
export function buildReportEscPos(doc: ReportEscPosDoc, cfg: ReceiptConfig): Uint8Array {
  const w = paperCols(cfg.paper);
  const line = (ch = "-") => ch.repeat(w);
  const two = (l: string, r: string) => {
    const left = l.slice(0, Math.max(0, w - r.length - 1));
    const space = Math.max(1, w - left.length - r.length);
    return left + " ".repeat(space) + r;
  };

  const b = new EscPosBuilder();
  b.init();
  b.align("center");
  b.bold(true).doubleHeight(true).line(doc.title).doubleHeight(false).bold(false);
  if (doc.subtitle) b.line(doc.subtitle);
  if (doc.meta) b.line(doc.meta);

  for (const sec of doc.sections) {
    b.align("left").line(line());
    if (sec.heading) b.bold(true).line(sec.heading.slice(0, w)).bold(false);

    const wide = (sec.columns?.length ?? 0) > 2;
    if (wide) {
      for (const row of sec.rows) {
        const [first, ...rest] = row.cells;
        b.bold(!!row.bold).line((first ?? "").slice(0, w)).bold(false);
        rest.forEach((val, i) => {
          const header = sec.columns![i + 1] ?? "";
          if (val) b.line(two(`  ${header}`, val));
        });
      }
    } else {
      for (const row of sec.rows) {
        const [label, value] = row.cells;
        b.bold(!!row.bold);
        b.line(two(label ?? "", value ?? ""));
        b.bold(false);
      }
    }
  }
  b.align("left").line(line());

  b.align("center");
  b.feed(3);
  b.cut(true);

  return b.build();
}

File: galaxyas-mobilepos/app/src/main/java/com/galaxyas/mobilepos/printer/EscPos.kt
package com.galaxyas.mobilepos.printer

import com.galaxyas.mobilepos.data.model.StockMovementBatchDetail
import com.galaxyas.mobilepos.data.model.TransactionDetail
import com.galaxyas.mobilepos.util.parseIso
import java.io.ByteArrayOutputStream
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.Normalizer
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Locale
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/**
 * Port line-for-line dari desktop/src/lib/escpos.ts. Byte hasilnya harus IDENTIK
 * dengan implementasi TS — dijaga oleh EscPosGoldenTest terhadap fixture di
 * app/src/test/resources/fixtures (berkas .hex). Jangan ubah tanpa memperbarui golden.
 */

private const val ESC = 0x1b
private const val GS = 0x1d

private val idLocale: Locale = Locale.forLanguageTag("id-ID")
private val idInteger: NumberFormat = NumberFormat.getIntegerInstance(idLocale)

/** "Rp157.500" — money() escpos.ts: "Rp" + Math.round(n).toLocaleString("id-ID"). */
private fun money(n: Double): String {
    val rounded = BigDecimal(n).setScale(0, RoundingMode.HALF_UP).toLong()
    return "Rp" + idInteger.format(rounded)
}

/** JS `${n}` template-literal: 3.0 -> "3", 2.5 -> "2.5" (dipakai baris item struk). */
private fun jsNum(n: Double): String {
    if (n == floor(n) && !n.isInfinite()) return n.toLong().toString()
    return BigDecimal.valueOf(n).stripTrailingZeros().toPlainString()
}

/** formatQty escpos.ts: Number.isInteger ? n.toString() : n.toFixed(2) (dipakai dok stok). */
private fun escFormatQty(n: Double): String =
    if (n == floor(n) && !n.isInfinite()) n.toLong().toString()
    else String.format(Locale.US, "%.2f", n)

/**
 * Padanan `new Date(iso).toLocaleString("id-ID")` di Node — "23/7/2026, 10.30.00"
 * (hari/bulan tanpa nol depan, jam 24h dengan pemisah titik). Pola EKSPLISIT,
 * bukan locale device, supaya byte struk identik lintas HP.
 */
private val idDateTimeFmt = SimpleDateFormat("d/M/yyyy, HH.mm.ss", Locale.US)
private fun idDateTime(iso: String): String {
    val d = parseIso(iso) ?: return iso
    return idDateTimeFmt.format(d)
}

/** Printer thermal umumnya CP437/ASCII: buang diakritik, sisanya non-ASCII -> '?'. */
private fun asciiBytes(s: String): List<Int> {
    val norm = Normalizer.normalize(s, Normalizer.Form.NFKD)
        .replace(Regex("[\\u0300-\\u036f]"), "")
    val out = ArrayList<Int>(norm.length)
    var i = 0
    while (i < norm.length) {
        val cp = norm.codePointAt(i)
        out.add(if (cp < 128) cp else 63)
        i += Character.charCount(cp)
    }
    return out
}

private class EscPosBuilder {
    private val bytes = ByteArrayOutputStream()

    fun raw(vararg b: Int): EscPosBuilder {
        for (x in b) bytes.write(x and 0xff)
        return this
    }

    fun text(s: String): EscPosBuilder {
        for (b in asciiBytes(s)) bytes.write(b and 0xff)
        return this
    }

    fun line(s: String = ""): EscPosBuilder = text(s).raw(0x0a)

    fun init(): EscPosBuilder = raw(ESC, 0x40)

    fun align(a: String): EscPosBuilder {
        val n = when (a) {
            "center" -> 1
            "right" -> 2
            else -> 0
        }
        return raw(ESC, 0x61, n)
    }

    fun bold(on: Boolean): EscPosBuilder = raw(ESC, 0x45, if (on) 1 else 0)

    fun doubleHeight(on: Boolean): EscPosBuilder = raw(GS, 0x21, if (on) 0x01 else 0x00)

    fun feed(lines: Int): EscPosBuilder = raw(ESC, 0x64, max(0, min(255, lines)))

    fun cut(partial: Boolean = true): EscPosBuilder = raw(GS, 0x56, if (partial) 1 else 0)

    fun build(): ByteArray = bytes.toByteArray()
}

private fun repeatChar(ch: Char, w: Int): String = ch.toString().repeat(w)

/** two(l, r): kolom kiri + kanan flush ke lebar w (padding spasi di tengah). */
private fun two(l: String, r: String, w: Int): String {
    val left = l.substring(0, max(0, min(l.length, w - r.length - 1)))
    val space = max(1, w - left.length - r.length)
    return left + " ".repeat(space) + r
}

private fun splitNonEmpty(s: String): List<String> =
    s.split("\n").map { it.trim() }.filter { it.isNotEmpty() }

/** Byte ESC/POS untuk struk transaksi, diakhiri feed + autocut. */
fun buildReceiptEscPos(detail: TransactionDetail, cfg: ReceiptConfig): ByteArray {
    val w = paperCols(cfg.paper)
    val show = cfg.show
    val b = EscPosBuilder()
    b.init()
    b.align("center")

    if (show.storeName) b.bold(true).doubleHeight(true).line(cfg.storeName).doubleHeight(false).bold(false)
    if (show.address && cfg.address.trim().isNotEmpty()) splitNonEmpty(cfg.address).forEach { b.line(it) }
    if (show.phone && cfg.phone.trim().isNotEmpty()) b.line(cfg.phone.trim())
    if (show.taxId && cfg.taxId.trim().isNotEmpty()) b.line("NPWP: ${cfg.taxId.trim()}")
    if (show.social) {
        val social = listOfNotNull(
            cfg.instagram.trim().ifEmpty { null }?.let { "IG: $it" },
            cfg.tiktok.trim().ifEmpty { null }?.let { "TikTok: $it" },
            cfg.whatsapp.trim().ifEmpty { null }?.let { "WA: $it" },
        ).joinToString(" · ")
        if (social.isNotEmpty()) b.line(social)
    }
    if (show.header) splitNonEmpty(cfg.header).forEach { b.line(it) }
    if (show.date) b.line(idDateTime(detail.created_at))
    if (show.invoiceNo) b.line(detail.invoice_no)

    if (show.items) {
        b.align("left").line(repeatChar('-', w))
        for (it in detail.items) {
            b.line(it.name.take(w))
            b.line(two("  ${jsNum(it.qty)} x ${money(it.price)}", money(it.price * it.qty), w))
            if (it.discount > 0) b.line(two("  Diskon", "-" + money(it.discount), w))
        }
    }

    val hasSummary = show.subtotal || show.discount || show.total || show.paymentMethod || show.change
    if (hasSummary) {
        b.align("left").line(repeatChar('-', w))
        if (show.subtotal) b.line(two("Subtotal", money(detail.subtotal), w))
        if (show.discount) b.line(two("Diskon", "-" + money(detail.discount), w))
        if (show.total) b.bold(true).line(two("TOTAL", money(detail.total), w)).bold(false)
        if (show.paymentMethod) {
            if (detail.payment_method == "Kombinasi") {
                b.line(two("Tunai", money(detail.paid_cash ?: 0.0), w))
                b.line(two("QRIS", money(detail.paid_qris ?: 0.0), w))
            } else {
                b.line(two(detail.payment_method, money(detail.paid), w))
            }
        }
        if (show.change) b.line(two("Kembali", money(detail.change), w))
        b.line(repeatChar('-', w))
    }

    if (show.footer && cfg.footer.trim().isNotEmpty()) {
        b.align("center")
        splitNonEmpty(cfg.footer).forEach { b.line(it) }
    }

    b.feed(3)
    b.cut(true)
    return b.build()
}

/** Byte ESC/POS untuk dokumen Item Masuk/Keluar. */
fun buildStockDocEscPos(detail: StockMovementBatchDetail, cfg: ReceiptConfig): ByteArray {
    val w = paperCols(cfg.paper)
    val show = cfg.show
    val verb = if (detail.kind == "in") "ITEM MASUK" else "ITEM KELUAR"
    val b = EscPosBuilder()
    b.init()
    b.align("center")

    if (show.storeName && cfg.storeName.trim().isNotEmpty()) {
        b.bold(true).doubleHeight(true).line(cfg.storeName).doubleHeight(false).bold(false)
    }
    b.bold(true).line(verb).bold(false)
    b.line(detail.no)
    b.line(idDateTime(detail.created_at))

    b.align("left").line(repeatChar('-', w))
    for (it in detail.items) {
        b.line(it.product_name.take(w))
        b.line(two("  Qty: ${escFormatQty(it.qty)}", "", w))
        if (!it.note.isNullOrEmpty()) b.line("  Ket: ${it.note}".take(w))
    }
    b.line(repeatChar('-', w))
    b.line(two("Total Item", detail.items.size.toString(), w))
    b.line(two("Total Qty", escFormatQty(detail.items.sumOf { it.qty }), w))
    if (!detail.note.isNullOrEmpty()) b.line("Catatan: ${detail.note}".take(w))
    if (!detail.user_id.isNullOrEmpty()) b.line("Oleh: ${detail.user_id}".take(w))
    b.line(repeatChar('-', w))

    if (show.footer && cfg.footer.trim().isNotEmpty()) {
        b.align("center")
        splitNonEmpty(cfg.footer).forEach { b.line(it) }
    }

    b.feed(2)
    b.cut(true)
    return b.build()
}

// ---------- Laporan sebagai struk ----------

data class ReportEscPosRow(val cells: List<String>, val bold: Boolean = false)
data class ReportEscPosSection(
    val heading: String? = null,
    val columns: List<String>? = null,
    val rows: List<ReportEscPosRow>,
)
data class ReportEscPosDoc(
    val title: String,
    val subtitle: String? = null,
    val meta: String? = null,
    val sections: List<ReportEscPosSection>,
)

fun buildReportEscPos(doc: ReportEscPosDoc, cfg: ReceiptConfig): ByteArray {
    val w = paperCols(cfg.paper)
    val b = EscPosBuilder()
    b.init()
    b.align("center")
    b.bold(true).doubleHeight(true).line(doc.title).doubleHeight(false).bold(false)
    doc.subtitle?.let { b.line(it) }
    doc.meta?.let { b.line(it) }

    for (sec in doc.sections) {
        b.align("left").line(repeatChar('-', w))
        sec.heading?.let { b.bold(true).line(it.take(w)).bold(false) }

        val wide = (sec.columns?.size ?: 0) > 2
        if (wide) {
            for (row in sec.rows) {
                val first = row.cells.firstOrNull() ?: ""
                b.bold(row.bold).line(first.take(w)).bold(false)
                row.cells.drop(1).forEachIndexed { i, value ->
                    val header = sec.columns!!.getOrNull(i + 1) ?: ""
                    if (value.isNotEmpty()) b.line(two("  $header", value, w))
                }
            }
        } else {
            for (row in sec.rows) {
                val label = row.cells.getOrNull(0) ?: ""
                val value = row.cells.getOrNull(1) ?: ""
                b.bold(row.bold)
                b.line(two(label, value, w))
                b.bold(false)
            }
        }
    }
    b.align("left").line(repeatChar('-', w))

    b.align("center")
    b.feed(3)
    b.cut(true)
    return b.build()
}
